### ISSUE TITLE
Synchronized complementary tracks of non-traditional media

### ISSUE THREAD

**Issue by ac-freeman** [moq-wg/moq-transport#640]
This is my first time jumping in on the MoQ working group, so please forgive my ignorance if I'm missing something in the draft or existing issues.

I'm working on a streaming system for non-traditional media known as [event-based video](https://en.wikipedia.org/wiki/Event_camera). It has a very high (and highly variable) data rate, and no image frames; pixels have independent, asynchronous samples that are sparse in time. 

Since the pixel samples are independent, I've explored splitting them up across several tracks. The client can then recombine the samples from any tracks that it's subscribed to. Since the samples are independent, this process doesn't rely on certain data to be decoded first, as with SVC. For a given segment of time, each track that we publish any data to will have roughly the _same_ amount of data. However, some tracks might not have _any_ data for that time segment, in the case that the overall data rate from the camera drops.

For low-latency applications, it's vital that the tracks are synchronized, so that each subscription for a given client is getting the latest data possible. If there's congestion when subscribing to a new track, however, the first group received from that track may be older than what the other tracks are sending. What mechanisms are there currently to avoid such a problem?

**Comment by afrind** [comment]
Welcome to MoQ and thanks for your interest.  #475 describes a similar problem.  There was a presentation at the last IETF meeting in November about potentially using timestamps for cross-track synchronization.

It's also possible to use subgroups to organize independent data within a single track.  The main differences between multiple tracks and a single track with multiple subgroups are:

1. A subscriber can independently subscribe and unsubscribe from tracks, but not subgroups
2. All objects in the same track share an ID space regardless of subgroup, but multiple tracks have independent ID spaces (though they can be theoretically coordinated).

**Comment by ac-freeman** [comment]
Thank you for the information. I definitely need to stick with a multiple-track approach, since the receiver performs rate-based adaptation through (un)subscribing to additional tracks. Timestamp-based synchronization would help, I think, but I suppose the use case is different than what's been discussed in that issue.

In this case, the client may have knowledge about the end-to-end latency, assuming some send timestamp is encoded in the object payload. The latency is below the target threshold, so it can receive more data. The client subscribes to one more track, but there's congestion at the relay then, so the subscription takes a while. Now the second track is out of sync. The congestion dies down, and the client issues a subscribe update to get things back in check.

In isolation, this isn't a big deal. But for my use case, tracks may be added and dropped with high frequency. This is for machine vision applications, though, so there's no concept of human QoE.

Another strategy may be to issue new SUBSCRIBEs with an AbsoluteStart set for some object in the future that hasn't been sent yet. But there would be no guarantee that the congestion will have recovered by then to the point that the tracks are resynchronized. It's also not clear to me if the protocol currently does or should define the behavior for requesting group/object IDs that are later than the newest data.

**Comment by ac-freeman** [comment]
> 2. All objects in the same track share an ID space regardless of subgroup

To confirm, does this mean that you cannot have two objects with the same ID in the same group, but different subgroups?

**Comment by afrind** [comment]
> It's also not clear to me if the protocol currently does or should define the behavior for requesting group/object IDs that are later than the newest data.

It does.  You can send a SUBSCRIBE with a start group/object in the future.

> To confirm, does this mean that you cannot have two objects with the same ID in the same group, but different subgroups?

Correct.  Track+Group+ID uniquely identifies an object.

On your overall problem, what does it mean for tracks to be out of sync?  Is it that track A is delivering data at time T and track B is delivering data from T - delta?  One way to help address this in MOQT is to set a delivery timeout in your subscription.  If the track B data is too old is will be skipped or reset, and the system should converge on delivering data that is approximately the same age, +/- the timeout value.

**Comment by ac-freeman** [comment]
> One way to help address this in MOQT is to set a delivery timeout in your subscription

Yes, that was the idea I had over the weekend. I'm thinking that for my purposes it might work with the subgroup approach plus the delivery timeout. Then you wouldn't have the overhead of (un)subscribing to several tracks when the bandwidth for the relay-client connection changes--just update the timeout delivery for the one track subscription.

**Comment by afrind** [comment]
@ac-freeman : is there a change to the draft you are seeking via this issue or is it ok to close now?  As I recall you were able to make things work via delivery timeouts.
